File: content/19-advanced-cp-patterns/kmp_search.py
"""
Problem: KMP Pattern Searching
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/longest-prefix-suffix2527/1
"""

def kmp_search(pat: str, txt: str) -> list[int]:
    """
    KMP search algorithm. Returns 0-based indices of pattern matches.
    Time Complexity: O(N + M)
    Space Complexity: O(M)
    """
    m = len(pat)
    n = len(txt)
    res = []
    
    if m == 0 or n == 0 or m > n:
        return []

    # Compute LPS (Longest Prefix Suffix) array
    lps = [0] * m
    length = 0 # length of the previous longest prefix suffix
    i = 1
    
    while i < m:
        if pat[i] == pat[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1

    # Search pattern
    i = 0 # index for txt
    j = 0 # index for pat
    while i < n:
        if pat[j] == txt[i]:
            i += 1
            j += 1
            
        if j == m:
            res.append(i - j)
            j = lps[j - 1]
        elif i < n and pat[j] != txt[i]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1
                
    return res

if __name__ == "__main__":
    assert kmp_search("ABABCABAB", "ABABDABABCABAB") == [5]
    assert kmp_search("TEST", "THIS IS A TEST TEXT") == [10]
    assert kmp_search("A", "AAAAA") == [0, 1, 2, 3, 4]
    print("Python: kmp_search passed tests!")

File: content/19-advanced-cp-patterns/kmp_search.js
/**
 * Problem: KMP Pattern Searching
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/longest-prefix-suffix2527/1
 */

function kmpSearch(pat, txt) {
    /**
     * KMP search algorithm.
     * Time Complexity: O(N + M)
     * Space Complexity: O(M)
     */
    const m = pat.length;
    const n = txt.length;
    const res = [];
    
    if (m === 0 || n === 0 || m > n) return [];
    
    const lps = new Array(m).fill(0);
    let len = 0;
    let i = 1;
    
    while (i < m) {
        if (pat[i] === pat[len]) {
            len++;
            lps[i] = len;
            i++;
        } else {
            if (len !== 0) {
                len = lps[len - 1];
            } else {
                lps[i] = 0;
                i++;
            }
        }
    }
    
    i = 0;
    let j = 0;
    while (i < n) {
        if (pat[j] === txt[i]) {
            i++;
            j++;
        }
        
        if (j === m) {
            res.push(i - j);
            j = lps[j - 1];
        } else if (i < n && pat[j] !== txt[i]) {
            if (j !== 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
    }
    
    return res;
}

if (require.main === module) {
    const assert = require('assert');
    assert.deepStrictEqual(kmpSearch("ABABCABAB", "ABABDABABCABAB"), [5]);
    assert.deepStrictEqual(kmpSearch("TEST", "THIS IS A TEST TEXT"), [10]);
    assert.deepStrictEqual(kmpSearch("A", "AAAAA"), [0, 1, 2, 3, 4]);
    console.log("JS: kmpSearch passed tests!");
}
